stmhal/uart.c

```diff
@@ -320,6 +320,22 @@ int uart_rx_char(pyb_uart_obj_t *self) {
             data = self->read_buf[self->read_buf_tail];
         }
         self->read_buf_tail = (self->read_buf_tail + 1) % self->read_buf_len;
+        if (__HAL_UART_GET_FLAG(&self->uart, UART_FLAG_RXNE) != RESET) { // May be stalled by flow ctrl: read UART
+            #if defined(MCU_SERIES_F7)
+            int data_dr = self->uart.Instance->RDR; // clears UART_FLAG_RXNE
+            #else
+            int data_dr = self->uart.Instance->DR; // clears UART_FLAG_RXNE
+            #endif
+            data_dr &= self->char_mask;
+            if (self->char_width == CHAR_WIDTH_9BIT) {
+                ((uint16_t*)self->read_buf)[self->read_buf_head] = data_dr;
+            } else {
+                self->read_buf[self->read_buf_head] = data_dr;
+            }
+            uint16_t next_head = (self->read_buf_head + 1) % self->read_buf_len;
+            self->read_buf_head = next_head;
```

Comment:
I think this entire chunk can go.  It's duplicated by the irq handler, and just removing lines 324-336 seems to make no difference because the following line (337) enables the IRQ which fires immediately and reads the pending char from DR.